core/parser.mly

```diff
@@ -299,102 +333,98 @@ declaration:
 
 nofun_declaration:
 | alien_block                                                  { $1 }
-| ALIEN VARIABLE STRING var COLON datatype SEMICOLON           { let {node = name; pos = name_pos} = $4 in
-                                                                 with_pos (`Foreign (make_untyped_binder name name_pos, name, $2, $3, datatype $6))
-                                                                           (pos()) }
+| ALIEN VARIABLE STRING var COLON datatype SEMICOLON           { with_pos $loc
+                                                                          (`Foreign (make_untyped_binder $4, $4.node, $2, $3, datatype $6)) }
 | fixity perhaps_uinteger op SEMICOLON                         { let assoc, set = $1 in
                                                                    set assoc (from_option default_fixity $2) ($3.node);
-                                                                   with_pos `Infix (pos()) }
-| tlvarbinding SEMICOLON                                       { let ({node=d;pos=dpos},p,l), pos = $1
-                                                                 in with_pos (`Val ([],
-                                                                                     (with_pos (`Variable (make_untyped_binder d dpos)) pos),p,l,None))
-                                                                              pos }
-| signature tlvarbinding SEMICOLON                             { annotate $1 (`Var $2) }
+                                                                   with_pos $loc `Infix }
+| tlvarbinding SEMICOLON                                       { let (bndr,p,l) = $1
+                                                                 in with_pos $loc($1) (`Val ([],
+                                                                    (with_pos $loc($1) (`Variable (make_untyped_binder bndr))), p, l, None)) }
+| signature tlvarbinding SEMICOLON                             { annotate $loc($1) $1 $loc($2) (`Var $2) }
 | typedecl SEMICOLON                                           { $1 }
 
 | links_module                                                 { $1 }
-| links_open                                                   { $1 }
+| links_open SEMICOLON                                         { $1 }
```

Comment:
I think this is fine, since other toplevel declaration forms are followed by semicolon so this is more uniform anyway.  

An alternative would be to require all "open"s to be at the beginning of the file with at least one normal declaration afterwards, but having to write `open Foo;` seems like a small price to pay for `open` being more first-class.